### ISSUE TITLE
new amalgkit subcommand for cross-species correlation analysis

### ISSUE THREAD

**Issue by kfuku52** [kfuku52/amalgkit#13]
![image](https://user-images.githubusercontent.com/19270515/91972027-635a0d00-ed1a-11ea-97f6-266d1108bc92.png)

**Comment by kfuku52** [comment]
Isn't it supported in the latest version?

**Comment by Hego-CCTB** [comment]
`cstmm` provides the cross species normalized counts, but the code for this exact plot is not in place. The code for this is in an R script separate from both `cstmm` and `curate`. It'd probably make the most sense to add this to `curate`.

There is a related discussion in here:
https://github.com/kfuku52/amalgkit/issues/11

**Comment by kfuku52** [comment]
Could you add this functionality as a separate subcommand? `curate` may be done in an array job per species, so individual jobs do not know whether the `curate` outputs are ready for this plot.

**Comment by Hego-CCTB** [comment]
yeah, I'm currently looking at the scripts again and making a new subcommand is probably the easiest way.

**Comment by Hego-CCTB** [comment]
I'm currently testing `amalgkit csca` (cross species correlation analysis), with a dataset of 11 species. The algorithm is very easily broken by 'N/A's in the dataset, as things like PCA or tSNE depend on NA free datasets. Currently trying to figure out where these 'N/A's are getting generated in my test set. 

When I get this to run on my test set, I'll push the update.

**Comment by Hego-CCTB** [comment]
So I finally got `amalgkit csca` to run on my test data. I had to remove 2 species and rerun Orthofinder to get a handful more Singlecopy orthogroups as well as adjust some of the gene name parsing. The script now runs without issue, but I still have this problem:
[Multisp.SVA.combined.pdf](https://github.com/kfuku52/amalgkit/files/6708145/Multisp.SVA.combined.pdf)

The differences between 'corrected' and 'uncorrected' are miniscule. 

The data ran through `cstmm` first, then `curate` with FPKM. So the result should be TMM-logFPKM-SVA normalized.

This is the rest of the output. 

[csca.zip](https://github.com/kfuku52/amalgkit/files/6708184/csca.zip)

**Comment by kfuku52** [comment]
Good progress! The result wouldn't be too bad in PCA. Why the heatmap looks so chaotic would be the presence of some outliers. For example, you may remove Helianthus if there is no good flower samples. I am also wondering if Helianthus leaf samples are contaminated with flowers (see heatmaps). Sometimes SRA metadata are mislabeled. Could you share the metadata table and species-wise `curate` plot for this species?

Ananas and Amborella may be removed too because these species have a strong species-specific feature, which couldn't be removed even by `curate` (see dendrograms).

**Comment by Hego-CCTB** [comment]
[Helianthus_annuus_curate_data.zip](https://github.com/kfuku52/amalgkit/files/6708330/Helianthus_annuus_curate_data.zip)

Sure, here you go!

To me, this looks fine, although all of the flower samples come from the same bioproject.

**Comment by Hego-CCTB** [comment]
I just noticed, that the plot in the bottom left says "Max expression (log TPM +1)". I double checked, this is hardcoded into the draw_exp_level_histogram() function in `curate`. I'll fix this to reflect the correct norm method with the next update.

**Comment by kfuku52** [comment]
I see two problems here in Helianthus.

1. SRR5803831 should have been removed by `curate`. This should be a bug.

2. SVA seems to over-correct some root samples:
SRR8302237
SRR8302249
SRR8302248
SRR8302250
SRR8302252
SRR8302253

Those samples show almost no correlation with the other root samples. They survived curate probably because Pearson r against other root samples (e.g., 0.02) are slightly higher than those against the other tissues (e.g., 0.01). To remove such samples, we should introduce a cutoff. The samples with within-tissue mean Pearson r < 0.2 would be OK to remove. For this, pearson r should be obtained without the same-tissue samples from the same BioProject, and without the SVA correction. And this step should be integrated into the existing `for` loop of outlier removal, to be able to remove the worst outlier(s) at a time.

**Comment by Hego-CCTB** [comment]
Good catch, this is indeed a problem!

Just so we are on the same page here:
- this would run separately from `check_within_tissue_correlation` and with raw logFPKM/logTPM values. The only thing that would change in terms of input between all iterations of outlier removal would be the number of SRR samples (the expression values would stay the same).
- Per iteration, this should only remove 1 outlier per tissue (instead of all r<0.2 samples at once) at maximum.

**Comment by kfuku52** [comment]
>This would run separately from check_within_tissue_correlation and with raw logFPKM/logTPM values.

No, it should be done within this while loop. I think it's easiest to modify `check_within_tissue_correlation` rather than making a new function.
https://github.com/kfuku52/amalgkit/blob/f5665c6af979aa27ded6317715aca3a40bf55755/amalgkit/transcriptome_curation.r#L809-L838

>Per iteration, this should only remove 1 outlier per tissue (instead of all r<0.2 samples at once) at maximum.

Here, 1 outlier means 1 SRA Run or same-tissue samples in 1 BioProject. We had a relevant discussion previously. Is it implemented and switchable with an option? If so, this step should follow the option.

**Comment by Hego-CCTB** [comment]
Gotcha!

> Here, 1 outlier means 1 SRA Run or same-tissue samples in 1 BioProject. We had a relevant discussion previously. Is it implemented and switchable with an option? If so, this step should follow the option.

This is not implemented yet, but as I'm modifying this function anyway, I might as well do it now.

**Comment by Hego-CCTB** [comment]
During testing I realized, that check_within_tissue_correlation() always worked on the original transformed fpkm/tpm data, instead of sva corrected ones. So I just calculate pearson r excluding same-tissue-and-same-bioproject samples and added a 0.2 threshold. I also included the option to only remove 1 outlier per same bioproject or same tissue.

In the above `Helianthus annuus` dataset, all flower samples come from the same bioproject, meaning I can't calculate pearson R with those same rules, as I would completely remove all flower samples from the set. How should I handle those cases?

As a side note, just introducing the 0.2 threshold on the pearson r INCLUDING same-tissue-same-bioproject samples, catches all the root samples you mentioned above already.

**Comment by kfuku52** [comment]
If there is only one BioProject in the tissue, we should obtain Pearson r only with those same-BioProject samples, so that we can remove bad samples that belong to a BioProject in which not all samples are bad.

>During testing I realized, that check_within_tissue_correlation() always worked on the original transformed fpkm/tpm data, instead of sva corrected ones.
Thank you for reporting. This problem seems to be inherited from my original script. I thought SVA was applied before the correlation check and described so in the NC paper; I'll prepare a corrigendum of the Method section.

**Comment by Hego-CCTB** [comment]
> Thank you for reporting. This problem seems to be inherited from my original script. I thought SVA was applied before the correlation check and described so in the NC paper; I'll prepare a corrigendum of the Method section.

SVA happens before the correlation check, it's just that it's stored in a different table than what is giving to the correlation check. `check_within_tissue_correlation()` could theoretically be fed with sva adjusted values too.

**Comment by kfuku52** [comment]
You can compare the performance but, since there were traces of trial and error in my original code, I think it would have been better for the animal data set to do correlation checks without SVA.

**Comment by Hego-CCTB** [comment]
Wait! Before the corrigendum, let me check the original script. There is this line in the outlier correction loop, which is currently commented out:
`# out = check_within_tissue_correlation(tc_sva, sra, dist_method, min_dif, selected_tissues) ; tc_sva `
This would actually feed the funciton with sva corrected values.

**Comment by Hego-CCTB** [comment]
> Wait! Before the corrigendum, let me check the original script. There is this line in the outlier correction loop, which is currently commented out:
> `# out = check_within_tissue_correlation(tc_sva, sra, dist_method, min_dif, selected_tissues) ; tc_sva `
> This would actually feed the funciton with sva corrected values.

So, this is comment is part of the original script too:
[transcriptome_curation.r.zip](https://github.com/kfuku52/amalgkit/files/6726672/transcriptome_curation.r.zip)

> You can compare the performance but, since there were traces of trial and error in my original code, I think it would have been better for the animal data set to do correlation checks without SVA.

Yeah, it seems like you had this thought back then too, seeing as the line was there originally.

**Comment by kfuku52** [comment]
So if SVA isn't really necessary for each round of iterative outlier removals, we can make it optional to plot SVA-corrected pdf (defaulted "no"), to speed up the iterations.

**Comment by Hego-CCTB** [comment]
This is already part of `curate` with the `--cleanup` option. Although this option is in need for an overhaul. I just noticed the help description is straight up wrong (0 means no intermediate SVA plots) and can be done with strtobool instead and yes/no instead. Also `cleanup` is the wrong choice of name, considering what it does. 

The way `--cleanup` works is, if 0 (default) only the very first SVA (before outlier removal) and the very last SVA (after outlier removal) plots are being calculated and printed. If 1, every iteration gets an SVA calculation.

I'll fix this with the next update.

**Comment by Hego-CCTB** [comment]
The above comment is now fixed. `csca` is now live in 0.6.0.0
https://github.com/kfuku52/amalgkit/issues/49
